bsp/Src/BSP_SPI.c

```diff
@@ -0,0 +1,196 @@
+#include "BSP_SPI.h"
+
+#define LOCAL_BUFFER_SIZE 128 // Modify if needed
+static uint8_t localTxMetaBuffer[LOCAL_BUFFER_SIZE];
+static StaticQueue_t localTxMetaStatic;
+static QueueHandle_t localTxMetaQueue;
+static uint8_t localRxMetaBuffer[LOCAL_BUFFER_SIZE];
+static StaticQueue_t localRxMetaStatic;
+static QueueHandle_t localRxMetaQueue;
+
+static QueueHandle_t* _transmitQueuePtr;
+static QueueHandle_t* _receiveQueuePtr;
+
+/**
+ * @brief Initialize the provided SPI port
+ * 
+ * @param spiHandle The SPI handle to configure and init
+ * @return HAL_StatusTypeDef 
+ */
+HAL_StatusTypeDef BSP_SPI_Init(SPI_HandleTypeDef* spiHandle, QueueHandle_t* txPtr, QueueHandle_t* rxPtr) {
+    HAL_StatusTypeDef stat;
+
+    // assign local pointers
+    // and create metadata queues
+    _transmitQueuePtr = txPtr;
+    _receiveQueuePtr = rxPtr;
+    localTxMetaQueue = xQueueCreateStatic(LOCAL_BUFFER_SIZE, sizeof(uint8_t), localTxMetaBuffer, &localTxMetaStatic);
+    localRxMetaQueue = xQueueCreateStatic(LOCAL_BUFFER_SIZE, sizeof(uint8_t), localRxMetaBuffer, &localRxMetaStatic);
+    
+    // We are only allowing SPI2/3
+    if (spiHandle->Instance == SPI2) {
+        if (__HAL_RCC_SPI2_IS_CLK_DISABLED()) {
+            __HAL_RCC_SPI2_CLK_ENABLE();
+        }
+    }
+    else if (spiHandle->Instance == SPI3) {
+        if (__HAL_RCC_SPI3_IS_CLK_DISABLED()) {
+            __HAL_RCC_SPI3_CLK_ENABLE();
+        }
+    } else {
+        return HAL_ERROR;
+    }
+    
+    stat = HAL_SPI_Init(spiHandle);
+
+    if (stat == HAL_OK) {
+        // Enable the interrupts and priorities
+        if (spiHandle->Instance == SPI2) {
+            HAL_NVIC_SetPriority(SPI2_IRQn, 0, 0);
+            HAL_NVIC_EnableIRQ(SPI2_IRQn);
+        } else if (spiHandle->Instance == SPI3) {
+            HAL_NVIC_SetPriority(SPI3_IRQn, 0, 0);
+            HAL_NVIC_EnableIRQ(SPI3_IRQn);
+        }
+        __HAL_SPI_ENABLE(spiHandle); // manually enable the peripheral
+
+        // Register our TX and RX interrupt callbacks
+        // These two calls are important! You must also ensure the corresponding peripheral's
+        // #define USE_HAL_<periph>_CALLBACKS is set to 1 in the stm32f4xx_hal_conf.h file.
+        HAL_SPI_RegisterCallback(spiHandle, HAL_SPI_TX_COMPLETE_CB_ID, HAL_SPI_TxCpltCallback);
+        HAL_SPI_RegisterCallback(spiHandle, HAL_SPI_RX_COMPLETE_CB_ID, HAL_SPI_RxCpltCallback);
+    }
+
+    return stat;
+}
+
+/**
+ * @brief Set SPI clock speed
+ * 
+ * @param spiHandle The SPI handle whose clock you're configuring
+ * @return HAL_StatusTypeDef 
+ */
+HAL_StatusTypeDef BSP_SPI_SetClock(SPI_HandleTypeDef* spiHandle, BSP_SPI_Speed speed) {
+    HAL_StatusTypeDef stat;
+
+    __HAL_SPI_DISABLE(spiHandle); // Disable the SPI periph while we set clock
+    spiHandle->Init.BaudRatePrescaler = _speedLut[speed];
+    stat = HAL_SPI_Init(spiHandle);
+    if (stat == HAL_OK) {
+        __HAL_SPI_ENABLE(spiHandle); // Re enable if all ok
+    }
+    return stat;
+}
```

Comment:
Can this logic just be in the init function (so we don't need to call multiple functions to make SPI work).